### PR TITLE
feat: implement improved highlights form with ai auto-summarise

### PR DESCRIPTION
**Pull request by OgDev-01**
## Description
This PR:
- Adds a new look to the highlights input form for desktop and mobile
- implements auto summarize for highlights 
- adds a new component to the design system
- updates the text area component to include typewriting mode
- improves the overall user experience for creating highlights on mobile and desktop devices

### Notes on summary generation
After testing the various `tone` available, I choose `informative` as it seems to generate a more in-depth context of what was done in the PR in fewer words. `formal` on the other hand was also nice, but it kind of generates almost same words as what was in the commit messages in the PR. I'm open to hearing thoughts on this  🍕 . 

 The generate summary icon is disabled by default and when a request is being made to get the summary from the API to prevent users from clicking when a request is in flight.

## Available tones
` "exciting" | "persuasive" | "informative" | "humorous" | "formal"`

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents
related #1289 

## Mobile & Desktop Screenshots/Recordings

image

https://github.com/open-sauced/insights/assets/62995161/f31ba88c-0728-4489-a98e-5bf37f07e313

image

image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by brandonroberts** [comment]
@OgDev-01 nice! Can we provide some feedback that the API request is in flight? Do we prevent them from clicking the suggestion button again while its happening?

**Comment by OgDev-01** [comment]
> @OgDev-01 nice! Can we provide some feedback that the API request is in flight? Do we prevent them from clicking the suggestion button again while its happening?

Opps!, totally missed that. updating shortly... Thanks!

**Comment by OgDev-01** [comment]
Update: I changed the `tone` back to `formal` and reduced `descriptionLength` to `400` as testing with my PR, the description generated were in markdown bullet lists and the length went a above the limit.

**Comment by bdougie** [comment]
Some feedback:

1. The generated response consistently comes back with summaries with more than 500 characters. Any chance we can keep it within the limit?
2. The blocking submission UI while waiting is ok, but my suggestion is a making the sparkle icon glow. 
3. I suggest we also generate a title. We can reuse the from the PR or create a prompt to provide a concise title based on the PR generated. This might need an api update to the endpoint to allow a system prompt in addition to the content prompt (git diff and commit messages) - I pitched that here: https://github.com/open-sauced/api/issues/203

## example prompt from - https://github.com/open-sauced/ai/pull/220
Screen Shot 2023-07-18 at 10 36 57 AM

## suggestion
1 & 2 should be addressed in this PR. 3 can just be using the PR title for now. We can do a follow up PR for title generation, unless you think it is approachable.

**Comment by bdougie** [comment]
@open-sauced/triage could use some extra eyes on this.

**Comment by bdougie** [comment]
Noticed the mobile pencil. Have you tested this with the sitegpt bubble and how they interact?

## mobile highlight pencil
Screen Shot 2023-07-19 at 8 59 11 AM

## sitegpt
Screen Shot 2023-07-19 at 9 00 20 AM

**Review by CBID2 — COMMENTED** [review]
Looks great @OgDev-01.

**Comment by OgDev-01** [comment]
> my suggestion is a making the sparkle icon glow.

Updates: I added a pulse animation to sparkle icon when a request is in flight. Tested with different PRs and none generates more than 500 words for me. @bdougie can you test with the updated changes again?

I changed the `descriptionLength` from `500` to `400` to fix 
> The generated response consistently comes back with summaries with more than 500 characters. Any chance we can keep it within the limit?

**Comment by OgDev-01** [comment]
here is a demo with the new animation effect


https://github.com/open-sauced/insights/assets/62995161/beefb251-5d29-424d-b302-f23a425c5a70

**Comment by takanome-dev** [comment]
> > Have you tested this with the sitegpt bubble and how they interact?
> 
> Not sure why, but for some reason, the sitegpt bubble doesn't seem to show up in dev and preview deploys on my own end.

@OgDev-01 yeah it's only visible on prod but you can turn it off to test this out https://github.com/open-sauced/insights/blob/ce126029e01247447be97912cebb6c1cbecca054/pages/_app.tsx#L63-L67

File: components/atoms/Textarea/text-area.tsx [hunk 1]
```diff
@@ -1,38 +1,64 @@
-import React, { ChangeEvent, useRef } from "react";
+import React, { ChangeEvent, useEffect, useRef } from "react";
 import clsx from "clsx";
 
 interface TextareaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
   defaultRow?: number;
   onChangeText?: (value: string) => void;
+  typewrite?: boolean;
```
Comment (brandonroberts) [inline components/atoms/Textarea/text-area.tsx:7]: I think this should be a separate `typewriter-textarea` atom or `molecule` and the `text-area` atom should remain as is. We only need this typewriter functionality in one place at the moment

File: lib/utils/generate-pr-highlight-summary.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        "Content-type": "application/json",
+        Authorization: `Bearer ${sessionToken}`,
+      },
+      body: JSON.stringify(payload),
+    });
+
+    if (res.ok) {
+      const data = await res.json();
+      return data.description as string;
+    }
```
Comment (brandonroberts) [inline lib/utils/generate-pr-highlight-summary.ts:29]: Needs a `null` return if no valid response also

File: lib/utils/github.ts [hunk 3]
```diff
@@ -55,4 +61,32 @@ const generateGhOgImage = (githubUrl: string): { isValid: boolean; url: string }
   }
 };
 
-export { getAvatarById, getAvatarByUsername, getProfileLink, getRepoIssuesLink, generateApiPrUrl, generateGhOgImage };
+const getPullRequestCommitMessageFromUrl = async (url: string): Promise<string[]> => {
+  const [, , , owner, repoName, , pullRequestNumber] = url.split("/");
+
+  const apiUrl = `https://api.github.com/repos/${owner}/${repoName}/pulls/${pullRequestNumber}/commits`;
```
Comment (brandonroberts) [inline lib/utils/github.ts:67]: This is going to get rate limited and potentially stop working without a token

**Review by brandonroberts — CHANGES_REQUESTED** [review]

**Comment by bdougie** [comment]
> Updates: I added a pulse animation to sparkle icon when a request is in flight. Tested with different PRs and none generates more than 500 words for me. bdougie can you test with the updated changes again?

Pulse looks great. 

re: 500 words, I was testing this PR https://github.com/open-sauced/ai/pull/220. But it seems fine now with the "formal" tone change. 

## system prompt

related to this issue on system prompts - https://github.com/open-sauced/api/issues/203

A system prompt is "return any GitHub issues numbers as markdown links." That addition to the prompt would make the generated summary include links instead of plain text issues numbers.

https://deploy-preview-1381--oss-insights.netlify.app/feed/154

Screen Shot 2023-07-20 at 1 13 49 PM

Note the link would need the markdown rendering from https://github.com/open-sauced/insights/pull/1376

**Comment by OgDev-01** [comment]
FYI, with the newly merged change from https://github.com/open-sauced/insights/pull/1396, running `npm run format` had a huge impact on the files changed

**Comment by OgDev-01** [comment]
image
Looking at the overlap in the above screenshot, i'm suggesting hiding the sitegpt action icon in the feeds page.

@bdougie what are your thoughts?

**Comment by OgDev-01** [comment]
Here is another suggestion
image

**Comment by bdougie** [comment]
> bdougie what are your thoughts?

Let's not go with both. I think that is the best option is hide sitegpt on the highlights page for mobile. I think in the future we should pin the input form to the top of the page on mobile. 

But to be clear just hide the SiteGPT bubble for now. 

Group 10

**Comment by brandonroberts** [comment]
@OgDev-01 why so many files changed?

**Comment by OgDev-01** [comment]
> @OgDev-01 why so many files changed?

> FYI, with the newly merged change from https://github.com/open-sauced/insights/pull/1396, running npm run format had a huge impact on the files changed

**Comment by brandonroberts** [comment]
Those changed files shouldn't be in your PR if you've merged in from beta though

**Comment by OgDev-01** [comment]
> Those changed files shouldn't be in your PR if you've merged in from beta though

I pulled from beta and got errors while i was trying to push, and then i ran `npm run format` and all those many files where formatted

**Review by brandonroberts — APPROVED** [review]
Functionality works. Extra files changed are from formatting

**Review by jpmcb — APPROVED** [review]
Looking good!!!! 👏🏼 

Screenshot 2023-07-21 at 3 39 51 PM

**Pull request merged**